### PR TITLE
fix: tolerate null referrers manifests on the consumer side

### LINKED ISSUE #398: FetchReferrersByApi throws NullReferenceException when registry returns "manifests": null
### Description

When listing referrers via the Referrers API, oras-dotnet crashes with a `NullReferenceException` if the registry returns `"manifests": null` instead of an empty array `"manifests": []`.

Per the [OCI Distribution Spec v1.1.1](https://github.com/opencontainers/distribution-spec/blob/v1.1.1/spec.md#listing-referrers), a registry with no referrers should return an empty `manifests` array. However, some registries (e.g. `dhi.io`) return `null` for the `manifests` field. This is technically a registry-side bug, but `oras` (Go) and `oras-go` tolerate it gracefully, so oras-dotnet should too.

### Root cause

`Oci.Index.Manifests` is declared as a non-nullable `required IList<Descriptor>`:

```csharp
[JsonPropertyName("manifests")]
public required IList<Descriptor> Manifests { get; set; }
```

When the JSON contains `"manifests": null`, `System.Text.Json` sets the property to `null`. Code in `Repository.FetchReferrersByApi` (and `PullReferrersIndexList`) then iterates `referrersIndex.Manifests`, throwing `NullReferenceException`.

### Why oras / oras-go are unaffected

oras-go decodes into `ocispec.Index`, where `Manifests` is a Go slice. A JSON `null` becomes a `nil` slice, and both `len(nil)` and `range nil` are safe no-ops. The `oras` CLI (`cp`, etc.) inherits this behavior — e.g. `cmd/oras/root/cp.go` does `if len(index.Manifests) == 0`, which is null-safe.

### Steps to reproduce

1. Point a copy/referrers operation at a repository on a registry that returns `"manifests": null` for a subject with no referrers (e.g. `dhi.io`).
2. Observe `NullReferenceException` during referrers enumeration.

### Proposed fix

Coalesce a `null` `manifests` value to an empty list in the `Index` model so it is never null after deserialization, mirroring Go's nil-slice semantics. Add tests covering deserialization of `"manifests": null` and an end-to-end `FetchReferrersByApi` test with a null-manifests response.

### Affected specs / references
- OCI Distribution Spec v1.1.1 – Listing Referrers
- Mirrors `oras-go` `registry/remote/repository.go` `referrersPageByAPI`

### PR DESCRIPTION
**Pull request by akashsinghal**
> **Credit:** This builds directly on @renshao's original fix in #399. It adopts
> the same goal and approach and addresses the additional review feedback with the
> consumer-side handling described below. @renshao is credited as a co-author on
> every commit in this PR.

## What this PR does

Fixes a `NullReferenceException` that occurs when a registry returns
`"manifests": null` for a subject with no referrers. Fixes #398.

## Spec rationale

The OCI image-index spec requires the `manifests` field to be **present**, but
does not require its value to be a non-null array. A registry with no referrers
is expected to return `[]`, but some (e.g. `dhi.io`) return `null`.
`Oci.Index.Manifests` is a non-nullable `required IList<Descriptor>`, so
System.Text.Json satisfies `required` (the key *is* present) yet sets the value
to `null` — and enumerating it later threw.

We keep `required` (the field MUST be present, so an entirely **absent**
`manifests` key is non-compliant and is intentionally still rejected at
deserialization) and tolerate only the spec-permissible **null value**.

## Why on the consumer side (supersedes #399)

#399 fixed this by coalescing `Oci.Index.Manifests` to an empty list on the
model. As @sajayantony noted, that stops an SDK author from holding or
**round-tripping** a `null` value, and re-serializing a `null` we accepted
should not silently become `[]`. So this PR keeps the `Index` model faithful to
the wire (null in → null out) and coalesces `null → empty` only where
deserialized registry responses are enumerated.

## Change

Coalesce `null → empty` at the three sites that enumerate a deserialized index's
manifests:

- `Repository.FetchReferrersByApi` — Referrers API path.
- `Repository.PullReferrersIndexList` — tag-schema fallback + referrers index
  update (`ManifestStore.ApplyReferrerChanges`).
- `FetchableExtensions.GetSuccessorsAsync` — copy/graph traversal (the default
  successor finder).

`Oci/Index.cs` is intentionally **unchanged**.

## Tests

- `Repository_ReferrersByApi_NullManifests_ReturnsEmpty` — Referrers API returns
  `manifests: null`; enumeration is empty (with and without an artifact-type filter).
- `ManifestStore_PullReferrersIndexList_NullManifests_ReturnsEmpty` — tag-schema
  index with `manifests: null` returns an empty list.
- `GetSuccessorsAsync_IndexManifestWithNullManifests_ReturnsEmpty` — copy/graph
  traversal of an index with `manifests: null` yields no successors.
- `Deserialize_NullManifests_PreservesNull` +
  `Serialize_NullManifests_WritesNullNotEmptyArray` — the model preserves `null`
  on round-trip (never rewrites to `[]`).

All 577 tests pass.

Supersedes #399 (with credit to @renshao as co-author).

### CODE REVIEW

File: src/OrasProject.Oras/Registry/Remote/Repository.cs [hunk 1]
```diff
@@ -606,7 +606,11 @@ internal async IAsyncEnumerable<Descriptor> FetchReferrersByApi(Descriptor descr
             // Set ReferrerState to Supported
             SetReferrersState(true);
 
-            var referrers = referrersIndex.Manifests;
+            // A spec-conformant registry returns an empty `manifests` array when there are
+            // no referrers, but some registries return `null`. Coalesce here so the null
+            // stays out of the model (SDK authors can still set/serialize null) while
+            // enumeration below is safe.
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply src/OrasProject.Oras/Registry/Remote/Repository.cs]: Good catch — reworded in f44bf99. It now describes this as local, enumeration-only coalescing and no longer implies the deserialized `Index` is mutated or that authoring a null `Manifests` is idiomatic (the property stays `required`).

[the remark this answers is not in the document]
Comment (akashsinghal) [reply src/OrasProject.Oras/Registry/Remote/Repository.cs]: Good catch — reworded in f44bf99. It now describes this as local, enumeration-only coalescing and no longer implies the deserialized `Index` is mutated or that authoring a null `Manifests` is idiomatic (the property stays `required`).

File: src/OrasProject.Oras/Registry/Remote/Repository.cs [hunk 2]
```diff
@@ -689,7 +693,10 @@ internal async IAsyncEnumerable<Descriptor> FetchReferrersByTagSchema(Descriptor
             var index = OciJsonSerializer.Deserialize<Index>(indexBytes)
                 ?? throw new JsonException(
                     $"Error deserializing index manifest for referrersTag {referrersTag}");
-            return (result.Descriptor, index.Manifests);
+            // Coalesce a non-conformant `null` manifests value to an empty list so callers
+            // (tag-schema fallback, referrers index update) can enumerate safely, while the
+            // model itself remains free to hold/round-trip null.
+            return (result.Descriptor, index.Manifests ?? ImmutableArray<Descriptor>.Empty);
         }
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply src/OrasProject.Oras/Registry/Remote/Repository.cs]: Agreed — this was already addressed in c4c9e82: `GetSuccessorsAsync` now does `descriptors.AddRange(index.Manifests ?? Array.Empty<Descriptor>())`, covered by a new `GetSuccessorsAsync_IndexManifestWithNullManifests_ReturnsEmpty` test. This review pass ran on the earlier commit (a329110).

File: tests/OrasProject.Oras.Tests/Serialization/ManifestSerializationTest.Index.cs [hunk 3]
```diff
@@ -259,6 +267,32 @@ public void Serialize_IndexEmptyManifests_ArrayPresent()
...
+    public void Deserialize_NullManifests_PreservesNull()
+    {
+        // The OCI spec does not forbid a null `manifests` value. Coalescing it on the
+        // model would stop an SDK author from producing/round-tripping null, so
+        // deserialization leaves it null; consumers guard against it instead.
+        var bytes = Encoding.UTF8.GetBytes(NullManifestsIndexJson);
+        var idx = OciJsonSerializer.Deserialize<OciIndex>(bytes)!;
+
+        Assert.Null(idx.Manifests);
+    }
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply tests/OrasProject.Oras.Tests/Serialization/ManifestSerializationTest.Index.cs:280]: This is an intentional design decision. The OCI image-index spec requires `manifests` to be **present** (MUST), so we keep it `required` — an entirely absent key is non-compliant and is still rejected at deserialization (verified: `JsonException` "missing required properties ... manifests"). The spec does not require the value to be a non-null array, so we tolerate a present-but-null value at the consumer boundaries only. Declaring the property `IList<Descriptor>?` would additionally relax the required-presence guarantee, which we don''t want. The round-trip tests document that we preserve (not normalize) the null value. See the PR description "Spec rationale" section.

File: src/OrasProject.Oras/Registry/Remote/Repository.cs [hunk 4]
```diff
@@ -689,7 +693,10 @@ internal async IAsyncEnumerable<Descriptor> FetchReferrersByTagSchema(Descriptor
             var index = OciJsonSerializer.Deserialize<Index>(indexBytes)
                 ?? throw new JsonException(
                     $"Error deserializing index manifest for referrersTag {referrersTag}");
-            return (result.Descriptor, index.Manifests);
+            // Coalesce a non-conformant `null` manifests value to an empty list so callers
+            // (tag-schema fallback, referrers index update) can enumerate safely, while the
+            // model itself remains free to hold/round-trip null.
+            return (result.Descriptor, index.Manifests ?? ImmutableArray<Descriptor>.Empty);
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply src/OrasProject.Oras/Registry/Remote/Repository.cs]: Fixed in f44bf99 — switched both the coalesced return and the pre-existing `catch (NotFoundException)` return to `Array.Empty<Descriptor>()`, and dropped the now-unused `System.Collections.Immutable` import.

File: src/OrasProject.Oras/Content/FetchableExtensions.cs [hunk 5]
```diff
@@ -64,7 +64,7 @@ public static async Task<IEnumerable<Descriptor>> GetSuccessorsAsync(this IFetch
                         // Note: Subject field only works for Oci Index Manifest
                         descriptors.Add(index.Subject);
                     }
-                    descriptors.AddRange(index.Manifests);
+                    descriptors.AddRange(index.Manifests ?? Array.Empty<Descriptor>());
```
Comment (sajayantony) [inline src/OrasProject.Oras/Content/FetchableExtensions.cs]: Given that this pattern exists for other properties maybe enabling an EnsureNotNull extension or someway to sanitize the other properties will avoid the same issue.

File: src/OrasProject.Oras/Registry/Remote/ManifestStore.cs [hunk 6]
```diff
@@ -290,7 +290,9 @@ private async Task ProcessReferrersAndPushIndex(Descriptor desc, Stream content,
                     return;
                 }
                 subject = imageManifest.Subject;
-                desc.ArtifactType = string.IsNullOrEmpty(imageManifest.ArtifactType) ? imageManifest.Config.MediaType : imageManifest.ArtifactType;
+                desc.ArtifactType = string.IsNullOrEmpty(imageManifest.ArtifactType)
+                    ? imageManifest.Config?.MediaType
```
Comment (akashsinghal) [inline src/OrasProject.Oras/Registry/Remote/ManifestStore.cs]: this change implies that it's possible for a config to not be present. In such a case, is the manifest even valid? even if deserialization to object type allows for null, from a spec POV are we allowed to accept this manifest if the config doesn't exist. if not, then we should be sanitizing this and throwing before we actually set the artifact type?

File: src/OrasProject.Oras/Content/FetchableExtensions.cs (hunk 5 above)
Comment (akashsinghal) [reply src/OrasProject.Oras/Content/FetchableExtensions.cs]: I think a good middle ground is to add a collection extension that will do the check and return an empty list or the original list if not empty/null. I've applied the change to work for `Layers` fields and `Manifests` field at the call sites (in each API) that interact with underlying object types where null is treated as a valid deserialized value.

The other remaining one with a similar class of null handling is the Config. The config is a required property in the spec but it is possible that the deserialized manifest provided has the field as null. oras-go automatically treats the descriptor as an initialized but empty descriptor on unmarshal. I think we need to consider doing the same. validation on the actual config descriptor (e.g digest, mediatype, size) should be performed on top of the empty descriptor separately. That validation is out of scope for this PR

File: src/OrasProject.Oras/Content/FetchableExtensions.cs [hunk 7]
```diff
@@ -48,8 +48,16 @@ public static async Task<IEnumerable<Descriptor>> GetSuccessorsAsync(this IFetch
...
                         descriptors.Add(manifest.Subject);
                     }
-                    descriptors.Add(manifest.Config);
-                    descriptors.AddRange(manifest.Layers);
+                    // A non-conformant manifest may send "config": null or "layers": null.
+                    // Treat a null as absent here (per-API) rather than propagating it into
+                    // the successor graph, where downstream consumers would throw. A
+                    // present-but-invalid config is left in place so it still fails through
+                    // existing validation. The deserialized manifest is left unchanged.
+                    if (manifest.Config is not null)
```
Comment (akashsinghal) [inline src/OrasProject.Oras/Content/FetchableExtensions.cs]: this really doesn't seem like the right behavior. if it's null, it's not valid manifest since config is required

File: src/OrasProject.Oras/Content/FetchableExtensions.cs (hunk 5 above)
Comment (akashsinghal) [reply src/OrasProject.Oras/Content/FetchableExtensions.cs]: Following up on my note above about mirroring oras-go (treating a null config as an initialized empty descriptor) — I pivoted away from that. The dividing line is **validity vs. treatment**:

- **`Config` — a validity question with one universal answer.** A manifest MUST have a valid config; a null/empty config is invalid *no matter which API consumes it*. Deserialization is where bytes become an object, so it's the right place to assert the object is well-formed: rejecting there loses nothing (invalid data isn't worth preserving) and every consumer gets the guarantee — none would want a different answer. So `Manifest` implements `IJsonOnDeserialized` and throws a clear `JsonException` on a null/invalid config. (STJ's `required` only checks key *presence*, so `"config": null` otherwise slips through.)

- **`Layers`/`Manifests` — a treatment question, where null is a *valid* representation.** An empty array is a valid spec state, so a wire `null` isn't malformed — just an unusual encoding of "no elements," and how you *treat* it can legitimately differ by consumer (an enumerator wants empty; an author inspecting/re-serializing may want the exact wire value preserved). Normalizing at deserialization would (a) be lossy — `null → empty` is irreversible, so you couldn't round-trip what the registry actually sent — and (b) impose one treatment on every consumer. So the model stays faithful (null stays null) and we normalize *locally at the point of use* via `NullToEmpty()`.

In short: deserialization is the right place to enforce **validity** (config), but the wrong place to bake in a **normalization** of a valid-but-unusual value (collections) — that stays a per-consumer choice.

Deeper descriptor validation (digest grammar, media-type IANA, size) stays out of scope here. (105f8ea)

File: src/OrasProject.Oras/Content/FetchableExtensions.cs [hunk 8]
```diff
@@ -49,7 +49,10 @@ public static async Task<IEnumerable<Descriptor>> GetSuccessorsAsync(this IFetch
                         descriptors.Add(manifest.Subject);
                     }
                     descriptors.Add(manifest.Config);
-                    descriptors.AddRange(manifest.Layers);
+                    // A non-conformant registry may send "layers": null instead of an empty
+                    // array (a valid, spec-permitted state); treat null as empty. Config is
+                    // guaranteed valid by Manifest deserialization validation.
+                    descriptors.AddRange(manifest.Layers.NullToEmpty());
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply src/OrasProject.Oras/Content/FetchableExtensions.cs:55]: this is not correct

**Review by sajayantony — APPROVED** [review]
